MaxText/inference/paged_attention.py

```diff
@@ -125,25 +126,69 @@ def paged_dot_product_attention_with_max_and_sum(self, query, key, value):
 
     return attn, local_max, local_sums
 
-  def paged_attention(
+  #TODO(rupliu): add sharding when SPMD is fully supported
+  def paged_attention_v2_prefill(
       self,
       query: Array,
       key_pages_var: nn.Variable,
       value_pages_var: nn.Variable,
       page_state: page_manager.PageState,
   ) -> Array:
-    """Apply Paged Attention.
-
-    Annotations:
-    b: batch_size
-    s: sequence length
-    n: query heads
+    """ Apply ragged input Paged Attention in prefill only.
+    """
+    k_p = jnp.permute_dims(key_pages_var.value, (1,2,0,3))
+    v_p = jnp.permute_dims(value_pages_var.value, (1,2,0,3))
+    c_q_l = jnp.array([0, page_state.sequence_lengths[0]]) #the accumulative real length of sequences
+    num_seqs = jnp.array([1]) 
+    query=query[0] #reduce [batch_size, max_num_tokens, num_kv_heads, head_dim] to [max_num_tokens, num_kv_heads, head_dim] since batch_size is always 1
+    result = paged_attention_kernel_v2.ragged_paged_attention(
+        q=query, 
+        k_pages=k_p, # [total_num_pages, page_size, num_kv_heads, head_dim]
+        v_pages=v_p,
+        kv_lens=jnp.array([query.shape[0]]), #max_prefill_length
+        cu_q_lens=c_q_l,
+        page_indices=page_state.page_map,
+        num_seqs=num_seqs,
+        #num_kv_pages_per_block=self.pages_per_compute_block, #I saw repeated response when enabled this
+    )
+    return jnp.expand_dims(result, axis=0)
 
-    k: kv_heads
-    x: num_pages
-    p: tokens_per_page
+  #TODO(rupliu): add sharding when SPMD is fully supported
+  def paged_attention_v2_decode(
+      self,
+      query: Array,
+      key_pages_var: nn.Variable,
+      value_pages_var: nn.Variable,
+      page_state: page_manager.PageState,
+  ) -> Array:
+    """ Apply ragged input Paged Attention in decode only.
+    """
+    batch_size = query.shape[0]
+    query = jnp.squeeze(query, axis=1) #reduce [batch_size, seq_len, n_kv_head, head_dim] to [batch_size, n_kv_head, head_dim]
+    k_p = jnp.permute_dims(key_pages_var.value, (1,2,0,3))
+    v_p = jnp.permute_dims(value_pages_var.value, (1,2,0,3))
+    c_q_l = jnp.arange(batch_size+1) #in decode, we expect one token per sequence
+    num_seqs = jnp.array([batch_size]) # real number of requests, set it to batch_size though it's not real_num_seq to avoid compilation
+    result = paged_attention_kernel_v2.ragged_paged_attention(
+        q=query, #[max_batched_num_tokens, num_kv_heads, head_dim]
+        k_pages=k_p, #[total_num_pages, page_size, num_kv_heads, head_dim]
+        v_pages=v_p, #[total_num_pages, page_size, num_kv_heads, head_dim]
+        kv_lens=page_state.sequence_lengths, #[max_num_seqs]
+        cu_q_lens=c_q_l, #[max_num_seqs+1]
+        page_indices=page_state.page_map, 
+        num_seqs=num_seqs,
+        num_kv_pages_per_block=self.pages_per_compute_block,
+    )
+    return jnp.expand_dims(result, axis=1)
 
-    d: kv_head_dim_size
+  def paged_attention_v1_decode(
```

Comment:
Is there any benefit to keeping the v1 code here?

Reply:
From XLA team's benchmarking, this kernel has 20% performance gain than v2 kernel in "decode only" tasks. So we could keep this in case we use different kernels for different stages

Reply:
Sounds good, maybe add that as a comment that v2 may not always be the best choice.

Reply:
will do